File: javascript/src/constantTime.js
'use strict';

const crypto = require('crypto');

/**
 * Timing-attack-resistant comparison. Accepts strings or Buffers.
 * @param {string|Buffer} a
 * @param {string|Buffer} b
 * @returns {boolean}
 */
function constantTimeEqual(a, b) {
  const bufA = Buffer.isBuffer(a) ? a : Buffer.from(String(a), 'utf8');
  const bufB = Buffer.isBuffer(b) ? b : Buffer.from(String(b), 'utf8');
  if (bufA.length !== bufB.length) {
    // Length mismatch is an intentional early exit -- see Go port's
    // ConstantTimeEqual doc comment for rationale.
    return false;
  }
  return crypto.timingSafeEqual(bufA, bufB);
}

module.exports = { constantTimeEqual };

File: java/src/main/java/io/github/securekit/ConstantTime.java
package io.github.securekit;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;

/**
 * Timing-attack-resistant comparison utility.
 */
public final class ConstantTime {

    private ConstantTime() {
    }

    public static boolean equals(byte[] a, byte[] b) {
        if (a.length != b.length) {
            // Length mismatch is an intentional early exit -- see the Go
            // port's ConstantTimeEqual doc comment for rationale.
            return false;
        }
        return MessageDigest.isEqual(a, b);
    }

    public static boolean equals(String a, String b) {
        return equals(a.getBytes(StandardCharsets.UTF_8), b.getBytes(StandardCharsets.UTF_8));
    }
}
